spec/intro.dd

```diff
@@ -67,6 +67,68 @@ $(OL
 )
 
 
+$(H2 Memory Model)
+
+    $(P A $(I byte) is the fundamental unit of storage. Each byte has 8 bits. Each byte has
+    a unique address. A $(I memory location) is a sequence of one or more bytes exactly large
+    enough to hold a scalar type. Multiple threads can access separate memory locations
+    without interference.
+    )
+
+    $(P Memory locations come in three groups:)
+
+    $(OL
+    $(LI $(I Thread local memory locations) are accessible from only one thread at a time.)
+    $(LI $(I Immutable memory locations) cannot be written to during its lifetime. Immutable
+    memory locations can be read from by multiple threads without synchronization.)
+    $(LI $(I Shared memory locations) is accessible from multiple threads.)
+    )
+
+    $(UNDEFINED_BEHAVIOR Allowing multiple threads to access a thread local memory
+    location results in undefined behavior.)
+
+    $(UNDEFINED_BEHAVIOR Writing to an immutable memory location during its lifetime
+    results in undefined behavior.)
+
+    $(UNDEFINED_BEHAVIOR Unless synchronized, writing to a shared memory location in
+    one thread while other threads read or write to it results in undefined behavior.)
+
+    $(P Execution of a single thread on thread local and immutable memory locations
+    is $(I sequentially consistent). This means the result of the operations
+    is the same as if it were executed in the order the operations appear in the program.
```

Comment:
This needs to be improved. "Sequentially consistent" is a term that applies if there are _multiple threads_ accessing _shared memory locations_ and it says that the observed result is as if the operations of the threads were _interleaved_, where the operations from each single thread are kept in order. Right now, this document doesn't say anything about shared memory locations.

Reply:
Sequentially consistent also applies even if there is only one thread. Of course we assume that, otherwise programs wouldn't work at all, but this sort of thing needs to be spelled out. (SC constrains what kinds of transformations the optimizer can apply.)